### DISCUSSION TITLE
Trend estimation, detrending and residual modeling

### DISCUSSION THREAD

**Discussion by krsnapaudel** [WUR-AI/AgML-CY-Bench#292] [Ideas]
Let's discuss approaches to trend estimation, detrending and residual modeling.

@mzachow, @ritviksahajpal, @mmeronijrc, @WestonAnderson.

**Comment by krsnapaudel** [comment]
Email thread I started:

1. First, the objective. We would like to have ML/DL models that predict not the full yields but the residuals from the trend. The trend will be location specific and year specific. This should remove the spatial and temporal dependencies in data points. After doing the detrending, I think we can assume the years are independent and data points are also independent. Correct me if I got this wrong.
2. Second, the order of leave-one-out vs detrending. Based on my understanding, we can assume years to be independent only if the trend has been removed from the yields. This means detrending has to come earlier than any leave-one-out validation. But LOO is needed to determine the training and test splits. So where do we start? How do we detrend? Or do we just assume years are independent from the start? NOTE this is what we are doing so far. But the baseline models don't do any detrending.
3. If we assume the years are independent, does it make sense to estimate trend? Is the assumption that years are independent saying there is no trend?
4. How do we estimate the trend? Current trend model takes the training data and fits a line through the data. The test year can be in the beginning, middle or end of the training years sequence. When the test year is in the middle, does this trend computation introduce information leakage? If yes, how do we fix this? Please check the [trend model implementation](https://github.com/BigDataWUR/AgML-CY-Bench/blob/main/cybench/models/trend_model.py). My current plan is to implement detrending using the current trend model implementation. I am not entirely comfortable with this. Hence the discussion.

@mmeronijrc:

1. I agree, if a “technological” yearly trend is present we can say that yearly data points are not independent. However, this seems to me a problem if the model used in the LOO actually models the trend using year before and after the left out year. In this case the trend component in the model should avoid information leakage and use trend info based on years before the left out (as it would happen in operations). If the model has not access to time (years of data points) it will able to recognize a “technological” trend (and position the left out year in the trend) only if the yield trend is connected to some other trend (long term increase of temperature, decrease of precipitation for example). Is it what you are concerned of? In this case one could also argue that it is not a time “technological” trend anymore but the model found a true link with temperature. Use of trending (in tuning or even after, in post-processing) is for sure useful to highlight the ability of the model to capture variations around the “business-as-usual”. That is, in the presence of a trend we would like to show error metrics saying how our model is performing beyond recognizing the trend.
2. not sure I get the point. You first remove “technological” and you have residuals (independent). Then you apply LOO.
3. NA
4. :My current understanding is that there is info leakage in LOO when you pass some info about the left out year to the model you are tuning. I don’t see this if you detrend first and the pass residuals to the model.

Final thought: I assume that in the work of Lily (working with long term climate projection) they might have the same problem. Meteo trend (global warming) coincides with time trend, if one removes the time trend one hides the meto trend.. What is their reasoning around that?

**Comment by krsnapaudel** [comment]
@WestonAnderson:
Once we’re dealing with trends the issue becomes quite subjective in the approach, but I broadly agree with Michele. My specific thoughts on your approach is below:

1. First, the objective. ... After doing the detrending, I think we can assume the years are independent and data points are also independent. Correct me if I got this wrong.
=> This is correct. It’s perfectly reasonable to assume independence from year-to-year after detrending.

2. Second, the order of leave-one-out vs detrending. ...
=> Detrending must come first. If a model requires an absolute yield value rather than a deviation, the most rigorous approach would be to calculate the trend using the past X years leading up to the withheld year, then project that trend forward to add to the model forecast anomaly. USDA, for example, uses a linear trend fit to the last 10 years in a moving window to forecast deviations from a trend. If the model is only forecasting the residuals then I would fit a linear trend using a moving window that ends on the year of interest. You only get unrealistic information leakage if you use a moving window centered on the year of interest (in which case the trend would draw information from future years).

3. If we assume the years are independent, does it make sense to estimate trend? Is the assumption that years are independent saying there is no trend?
=> Years are independent only after detrending. After detrending it’s probably reasonable to assume independence if you’re using a linear trend fit to a moving 10 year window even if there is some small degree of interannual autocorrelation remaining in the residuals.

4. How do we estimate the trend? 
=> The only point of caution I would add is that in some regions a linear trend becomes inappropriate if it is fit through ~30 years of data. The best way to tell is to plot the data and the trend.

I’ve not commented on Michele’s point about technological vs meteorological trends, but it’s a good point and something to consider.

**Comment by krsnapaudel** [comment]
@ritviksahajpal 

I agree with the comments from Weston and Michele. As a practical matter, given that we might only have crop yield data from 2000 onwards for several countries (or covariate data like MODIS), how about this approach for computing trends (based on what Weston outlined), consider a region with crop yield data from 2000 - 2023:
1. For years with at least ten years of historical data, e.g. 2012, use the ten years of historical data for computing trends, in this case from 2002-2011
2. For years with less than ten years of historical yield data, e.g., 2001, use the ten years of future yield data, e.g., 2002-2011, for computing trends.
3. Do not detrend if at least ten years of crop yield data (either historic or future) are not available for detrending for ALL years in the dataset. This would mean we ignore regions with less than twenty years of yield data. Alternatively, for these regions, we can do a sliding window-based validation, but that would make intercomparison challenging.

**Comment by krsnapaudel** [comment]
@sooranaresh:
1. The detrending is done to discount the technology based improvements in yield so that we can account for changes in yield due to weather of that season in which crop is grown. Hence if we take detrend data, we can still forecast the weather effects on yield as relative change over baseline.
2. If we do not do detrend, then the data will have trends due to technology and weather effects and may be useful to forecast in near-term assuming that the yield change due to technology is still at the same rate.

3. Year per se may not be important, if its characteristics in terms of weather are captured in the model.

**Comment by krsnapaudel** [comment]
I would like to include some machine learning related requirements.

1. We have to do the training-test set split before doing any detrending if we delegate detrending to modeling. This is because the test set is considered unseen data and may not be available during training. Weston mentioned that detrending must come first. We cannot do detrending first without knowing what the training dataset is. Any trend estimation and detrending must use the training data only.
2. Any modeling decision left to the modeler is tied to the training set. The modeler needs a training dataset to build and optimize a model. This also implies that we need the training and test splits before delegating trend estimation and detrending to the modeler. After passing the training dataset to the model or modeler, we cannot enforce or restrict them to use data earlier than the test year. In a LOO split, the test year can be in the middle of training years. The modelers will be free to do what they please with all of the training data.

**Comment by krsnapaudel** [comment]
@mmeronijrc:

If our final objective is to operationally forecast the absolute yield, I guess we should not detrend the yield in advance and let the modeler to decide if to: 1) build a trend model (say linear for now) and a residual model (the ML model), 2) incorporate both trend and residual in a single ML model (in various ways).
With both options the handling of the trend should be done in a way that avoid info leakage (do not use data earlier than test year, or the proposal of Ritvik).
It is true that years are not independent in the presence of a technological time trend. Is time independence really needed for LOO if we make sure that there is no info leakage?

@AlexRuane:
Just a few quick notes to add to this useful discussion:
* Even if we are aiming for absolute yields it may make sense to forecast yield anomalies and then convert them back to absolute yields using mean and variance corrections.
* In AgMIP GGCMI phase 1 (historical intercomparison) we looked at a number of de-trending techniques (linear, non-linear, moving windows, etc.).  The general answer was that no single approach was always the best, but the moving window tended to be the most useful.  Of course, a major drought can serve to depress the surrounding window mean in a way that somewhat suppresses apparent variability, but this is likely a necessary bias that we can acknowledge.  In many parts of the world even a 5-year moving window may be long compared to the pace of technological changes, so we are always going to introduce a little bias in this approach.
* On year-by-year independence – I think this is an important concept and allows us to have more statistically-plausible sequences to investigate.  It’s important to note that there are some carry-over effects in soil moisture and soil nutrient pools (particularly in less intensive systems, as shown by AgMIP Low-Input Systems Model Intercomparison Team), but this effect is likely small compared to the statistical benefit of treating systems as effectively independent.

**Comment by krsnapaudel** [comment]
I would say our final goal is to forecast yields. The method could estimate trend and forecast residuals. Then trend estimation and detrending must be a modeling choice.

**Comment by krsnapaudel** [comment]
@WestonAnderson:
I think Michele's first sentence is important. If we are aiming to forecast absolute yields, we should not detrend. If, however, we are aiming to forecast yield anomalies relative to a trend we will define our predictand dataset as the yield anomalies (not the absolute yields) and we will detrend prior to creating the test and train split. This is acceptable if we are defining the yield anomaly dataset itself as our target data rather than the absolute yields. While user demands for forecasts of absolute yields are far more common, there are some instances in which users want only forecasts of percent yield anomalies. We just need to be clear about what we are aiming to do.

**Comment by krsnapaudel** [comment]
@mzachow:

Regardless of the procedure for our benchmark dataset (where we need a protocol to publish ground truth residuals), I would like to hear your opinions on the following:

**Background**: Absolute in year `yield` can be described by $yield (y, x_y) = t(y) + \epsilon(x_y)$, where $t(y)$ is the trend model evaluated at the current year and $\epsilon(x_y)$ is the residual model, that depends on this year’s growth conditions $x_y$ (e.g. weather). In theory, the better we are able to dissect the trend from the residuals, the better our residual model will perform that only uses season-specific predictors. Hence, our goal should be to remove the trend as cleanly as possible if we want to build the best possible residual model. The problem I am seeing is the additional processing step that our target data undergoes. The ground truth of absolute yield data that we work with is already a somewhat flawed estimation, because nobody counted every single grain at harvest. Through de-trending, we may introduce an additional source of uncertainty to our target data, which may impact the ability of developing a residual model.

**My question**: I am wondering about the drawbacks of simply introducing the year as an additional quantity to the feature vector and let the ML model fit the entire function $yield (y, x_y)$? Wouldn’t this save us a lot of headache in terms of data leakage, trend model development, IID assumptions and validation procedure?

What did I miss? Curious to hear your thoughts!

**Comment by krsnapaudel** [comment]
From my point of view, LOO and providing year information to the model introduces information leakage when the test year is somewhere in the middle. For all purposes, the trend model has to deal with missing years. So the trend model will have no way to limiting trend estimation using data earlier than a test year. Using data from before and after the test year imho gives clues about the test year itself, hence the information leakage.

**Comment by krsnapaudel** [comment]
@mzachow

I think I see your point and have to agree!

Given the little practical relevance (?) of a pure residual model, I wouldn’t worry about info leakage to create the residuals and just fit a slope through the entire series.

As mentioned on GitHub: Isn’t this just a toy/model problem to see what are the limitations of residual model performance and to push research in the field, given the residuals are known?

If we de-trend without info leakage, our residuals may be flawed because they still carry a trend component.

In that case, how would I know if my residual model with a new proposed architecture performs poorly because of the changes I introduced or because the ground truth data is biased?

Maybe a bit exaggerated, but for me this translates into the following message: “Use CY-Bench data to develop a residual model. Predictors we have and we have something that we may think could be the residuals.”

Not sure if this is the right approach. Maybe I am missing the overall objective of publishing the benchmark data? What about the practical use of a pure residual model?

**Comment by wannh2** [comment]
@krsnapaudel @AlexRuane:

Just a note to add in this interesting discussion. 

I agree with Alex's a few points in detrending techniques and anomaly suggested. The properties of residuals are critical in any modeling approach, including the use of ML methods. Regarding the time series window, using a 10-year period might be a littl bit risky due to yield variations influenced by climate drivers (e.g., extended dry spells or significant anomalies occurring at the beginning or end of the window).  

I would like to share our recent two papers for your reference if interested, both of which used detrending (a kind removal of tech improvement): 

corn paper:  https://www.sciencedirect.com/science/article/pii/S016819232400282X
wheat paper: https://www.science.org/doi/10.1126/sciadv.ado6864

**Reply by krsnapaudel** [reply]
@wannh2: Thanks for sharing the paper. Can you share the code used to detrend? It's important to see the source of truth :)

**Comment by krsnapaudel** [comment]
@WestonAnderson on why we need a residual model:
Residual model: FEWS NET is interested in a residual model for two reasons: 1. They are looking to isolate the effects of climate from other impacts. They use other sources of information to develop expectations about how socioeconomic factors influence yields. 2. They apply a ‘deviation from baseline’, modeling approach in a wider food security modeling framework (household economy approach). This modeling effort would require translating even absolute yield forecast into a deviation from some definition of normal.

The one point I’d make about the trend vs residuals is that we again need to clearly define our interest. If we are specifically interested in the influence of weather on yields, we need to recognize that for an absolute yield model it will be impossible to separate model skill related to the accurate prediction of a non-climate trend from the skill related to predicting the influence of climate. The best model may not include any climate variables whatsoever. If we are simply interested in the best model, that is okay

**Comment by krsnapaudel** [comment]
Based on the discussion so far, questions and considerations:
1. What are the labels? At the moment, labels are yield statistics published at subnational level, curated and harmonized by organizations like JRC and FEWSNET.
2. Should we change the labels? If we want to detrend before any modeling, we need to estimate trend and remove the trend from the labels. In general, I am not enthusiastic about changing the labels in the dataset.
3. Let's say we decide to keep yields as labels. How do we split the data for validation and final evaluation? Can we still assume independence of years and use leave-one-out or cross-validation?
4. Let's say we decide to detrend first. Who is going to provide a trend estimation routine we can agree on? What are the steps for a reasonable trend estimation method? See my attempt [here](https://github.com/BigDataWUR/AgML-CY-Bench/issues/82).
5. Going back to our objective(s).
  a. Our objective is to produce yield forecasts. We can do that using yields as labels as well as residuals as labels. After forecasting residuals, we can add the trend back. Many of us have mentioned this.
  b. If we want to separate the contributions of trend and different predictors, we can do that through [interpretable machine learning](https://christophm.github.io/interpretable-ml-book/) methods as well. We don't need to enforce a trend estimation and detrending step on the labels. Feature attribution methods provide an estimate of which inputs contributed how much to the yield forecasts.
  c. A third objective is to evaluate on years with extreme yield anomalies. How do we define this set of years? Do we already have a list? If we do, then we can have a challenge test set including such years. We don't need to do leave-one-out for this.
6. All of the above can also be done with a time-based training and test split, with the most recent years in the test set. My understanding is that, due to climate change, yield anomalies are more likely in recent years than in the past. Similarly, the real-world application of yield forecasting is also to forecast for coming year or years. Why not make the models for this particular purpose?

**Comment by krsnapaudel** [comment]
Based on the NeurIPS reviews and the discussion so far, here is the best we can do:
1. We don't modify labels. So no detrending before splitting data.
2. Trend model will not do any trend estimation during training. Since trend is estimated using label data, trend estimation can be done per test year during inference. At the point we can decide the best strategy that avoids information leakage. At the moment, the best strategy is to extrapolate forward from previous years or backward from later years. In both cases we identify the optimal trend window.
3. If the data is insufficient to estimate trend or the trend is insignificant, we return the average.

See #307 for more details.
